### PR TITLE
Add performace diff benchmarks

### PR DESCRIPTION
**Pull request by soulomoon**
add something like this in benchmark result to easily view the performance diffferent like what we did for the svg.
```
Configuration, name, success, samples, startup, setup, userT, delayedT, 1stBuildT, avgPerRespT, totalT, rulesBuilt, rulesChanged, rulesVisited, rulesTotal, ruleEdges, ghcRebuilds, maxResidency, allocatedBytes
All, edit-header, True,0.00%,2.21%,Nah,-98.75%,Nah,67.50%, NaN,67.50%,-1.89%,-1.81%,-1.80%,-1.24%,0.39%,1650.00%,169.44%,49.89%
All, edit, True,0.00%,0.56%,Nah,3.66%,Nah,3.66%, NaN,3.66%,1.10%,1.10%,1.10%,0.00%,0.00%,0.00%,72.22%,-0.18%
All, hover, True,0.00%,0.57%,Nah,-43.57%,Nah,-43.57%, NaN,-43.57%,0.13%,0.13%,0.13%,1.07%,-0.39%,-94.29%,-61.50%,-33.32%
All, semanticTokens, True,0.00%,1.10%,Nah,2.22%,Nah,2.22%, NaN,2.22%,0.00%,0.00%,0.00%,0.00%,0.00%,0.00%,90.43%,0.46%
All, hover after edit, True,0.00%,5.88%,Nah,-1.22%,Nah,-1.22%, NaN,-1.22%,-1.55%,-1.55%,-1.55%,0.00%,0.00%,0.00%,118.07%,0.71%
All, getDefinition, True,0.00%,1.11%,Nah,8.57%,Nah,7.04%, NaN,7.04%,-0.71%,-0.71%,-0.71%,0.00%,0.00%,0.00%,33.33%,1.06%
All, getDefinition after edit, True,0.00%,-1.68%,Nah,10.53%,Nah,9.09%, NaN,9.09%,2.45%,2.45%,2.45%,0.00%,0.00%,0.00%,37.14%,-1.01%
All, completions, True,0.00%,0.56%,Nah,5.68%,Nah,5.68%, NaN,5.68%,0.31%,0.31%,0.31%,0.00%,0.00%,0.00%,0.00%,-0.48%
All, completions after edit, True,0.00%,9.20%,Nah,-3.03%,Nah,-3.03%, NaN,-3.03%,-1.37%,-1.37%,-1.37%,0.00%,0.00%,0.00%,-25.25%,0.54%
All, code actions, True,0.00%,-6.67%,Nah,-18.28%,0.00%,-18.09%, NaN,-18.09%,0.48%,0.48%,0.48%,0.04%,0.00%,0.00%,98.89%,-0.95%
All, code actions after edit, True,0.00%,-2.13%,Nah,-9.18%,Nah,-9.18%, NaN,-9.18%,-2.67%,-2.67%,-2.67%,0.00%,0.00%,0.00%,79.25%,0.37%
All, code actions after cradle edit, False,0.00%,2.30%,Nah,Nah,Nah,Nah, NaN,23.77%,-0.42%,-0.46%,-0.33%,-1.05%,0.39%,1650.00%,119.44%,44.27%
All, documentSymbols after edit, True,0.00%,-13.40%,Nah,50.00%,-47.48%,-44.14%, NaN,-44.14%,1.91%,1.91%,1.90%,1.07%,-0.39%,-94.29%,-17.37%,-32.40%
All, hole fit suggestions, True,0.00%,0.00%,Nah,1.15%,Nah,1.15%, NaN,1.15%,-0.93%,-0.93%,-0.93%,0.00%,0.00%,0.00%,101.92%,0.49%
All, eval execute single-line code lens, True,0.00%,-1.63%,Nah,-15.02%,Nah,-15.02%, NaN,-15.02%,10.00%,0.00%,12.29%,0.00%,0.00%,0.00%,0.88%,0.58%
All, eval execute multi-line code lens, True,0.00%,-1.60%,Nah,1.98%,Nah,1.98%, NaN,1.98%,0.00%,0.00%,-0.30%,0.00%,0.00%,0.00%,-0.84%,0.11%

```

### CODE REVIEW

**Comment by soulomoon** [comment]
I think it is ready now, https://github.com/haskell/haskell-language-server/actions/runs/8926967770/job/24520033550?pr=4203. Display result is showing percentage in addition

**Comment by michaelpj** [comment]
Could we emit some more information in the job to give some context for what these mean? It's not clear necessarily from looking at it. Also, are those numbers real? Do we really get random 1429% changes on stuff?

**Review by michaelpj — APPROVED** [review]
I haven't reviewed the code really, but maybe it needs some more comments? :)

File: shake-bench/src/Development/Benchmark/Rules.hs [hunk 1]
```diff
@@ -384,69 +391,89 @@ parseMaxResidencyAndAllocations input =
...
+diffItem (Mem x) (Mem y) = (Mem x, Just $ fromIntegral x / fromIntegral y)
+diffItem (Time x) (Time y) = (Time x, if y == 0 then Nothing else Just $ x / y)
+diffItem (ItemString x) (ItemString y) = (ItemString x, Nothing)
+diffItem _ _ = (ItemString "no match", Nothing)
+
+data Item = Mem Int | Time Double | ItemString String
+  deriving (Show)
 
-    writeFileChanged out $ unlines $ header' : concat results'
+-- split on ','
```
Comment (jhrcek) [inline shake-bench/src/Development/Benchmark/Rules.hs:462]:
```suggestion
```

File: shake-bench/src/Development/Benchmark/Rules.hs [hunk 2]
```diff
@@ -384,69 +391,89 @@ parseMaxResidencyAndAllocations input =
...
+parseLine :: String -> [Item]
+parseLine = map f . splitOn ","
+  where
+    f x
+      | "MB" `isSuffixOf` x = Mem $ read $ reverse $ drop 2 $ reverse x
+      --   is is double
```
Comment (jhrcek) [inline shake-bench/src/Development/Benchmark/Rules.hs:468]:
```suggestion
      --   is it double
```

File: shake-bench/src/Development/Benchmark/Rules.hs [hunk 3]
```diff
@@ -384,69 +391,89 @@ parseMaxResidencyAndAllocations input =
...
+convertToDiffResults :: String -> String -> String
+convertToDiffResults line baseLine = intercalate "," diffResults
+        where items = parseLine line
+              baseItems = parseLine baseLine
+              diffItems = zipWith diffItem items baseItems
+              diffResults = map showItemDiffResult diffItems
+
+showItemDiffResult ::  (Item, Maybe Double) -> String
+showItemDiffResult (ItemString x, _) = x
+showItemDiffResult (_, Nothing)      = "Nah"
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply shake-bench/src/Development/Benchmark/Rules.hs:449]: yes, NA

File: shake-bench/src/Development/Benchmark/Rules.hs [hunk 4]
```diff
@@ -384,69 +391,89 @@ parseMaxResidencyAndAllocations input =
...
-    writeFileChanged out $ unlines $ header' : concat results'
+-- split on ','
+parseLine :: String -> [Item]
+parseLine = map f . splitOn ","
+  where
+    f x
+      | "MB" `isSuffixOf` x = Mem $ read $ reverse $ drop 2 $ reverse x
+      --   is is double
+      | any isAlpha x = ItemString x
+      | otherwise = Time $ read x
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply shake-bench/src/Development/Benchmark/Rules.hs:470]: good point

**Comment by soulomoon** [comment]
> Do we really get random 1429% changes on stuff?
Yes, extream case could happen for small number division, like `3/0.01` : (

File: shake-bench/src/Development/Benchmark/Rules.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (soulomoon) [reply shake-bench/src/Development/Benchmark/Rules.hs]: done

File: shake-bench/src/Development/Benchmark/Rules.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (soulomoon) [reply shake-bench/src/Development/Benchmark/Rules.hs]: fixed

**Comment by soulomoon** [comment]
I added a header 
now we have
```
Performance Diff(comparing to its previous Version):
Version    Configuration   name                                 success   samples   startup   setup   userT    delayedT   1stBuildT   avgPerRespT   totalT   rulesBuilt   rulesChanged   rulesVisited   rulesTotal   ruleEdges   ghcRebuilds   maxResidency   allocatedBytes
upstream   All             edit-header                          True     0.00%     0.00%     NA      0.00%    0.00%      0.00%       0.00%         0.00%    0.00%        0.00%          0.00%          0.00%        0.00%       0.00%         0.00%          0.00%
HEAD       All             edit-header                          True     0.00%     -7.81%    NA      36.74%   -89.41%    5.23%       7.14%         7.11%    -11.11%      -6.25%         -0.37%         0.01%        0.00%       0.00%         1.20%          1.63%
```

**Pull request merged**